File: Java/ValidParanthesis.java
package Java;

import java.util.HashMap;
import java.util.Stack;
//The logic here is in the loop, first we check if the element is opening bracket if it is then we add it to stack expecting the closing bracket, then if we come across any closing brackets in string we check if the current top element in the stack is its matching opening bracket if not , then the string contains a invalid paranthesis and it must return false.If the loop gets over matching and popping elements from the stack then the stack must be empty
/**
 * ValidParanthesis
 */
public class ValidParanthesis {
public static void main(String[] args) {
    String s = "{[]}";
    System.out.println(valid(s));
    System.out.println(optimized_sln(s));

}
static boolean valid(String s){
    Stack<Character> stack = new Stack<>();
    char[] arr = s.toCharArray();
    for (char p : arr) {
        if(p == '{' || p == '(' || p == '[') stack.push(p);
        else if(p == ')' && !stack.isEmpty() && stack.peek() == '(') stack.pop();
        else if(p == '}' && !stack.isEmpty() && stack.peek() == '{') stack.pop();
        else if(p == ']' && !stack.isEmpty() && stack.peek() == '[') stack.pop();
        else return false;
    }
    return stack.isEmpty();
}
static boolean optimized_sln(String s){
    Stack<Character> stack = new Stack<>();
    HashMap<Character,Character> map = new HashMap<>();
    char[] arr = s.toCharArray();

    for (char p : arr) {
        if(map.containsKey(p)) stack.push(p);
        else if(map.containsValue(p)){
            if(stack.isEmpty() || stack.peek() != p) return false;
        }
    }
    return stack.isEmpty();
}
}

File: Python/Valid_Paranthesis.py
s = "()[]"
def valid_paranthesis(s) -> bool:
    arr = []
    for i in s:
        if(i == '{' or i == '(' or i == '['):
            arr.append(i)
        elif(i == '}' and len(arr) != 0 and arr[-1] == '{'):
            arr.pop()
        elif(i == ')' and len(arr) != 0 and arr[-1] == '('):
            arr.pop()
        elif(i == ']' and len(arr) != 0 and arr[-1] == '['):
            arr.pop()
        else:
            return False
    return len(arr) == 0
print(valid_paranthesis(s))
def more_optimized(s) -> bool:
    allowed = {'{':'}','(':')','[':']'}
    arr = []
    for p in s:
        if p in allowed:
            arr.append(p)
        elif p in allowed.values():
            if not arr or allowed[arr.pop()] != p:
                return False
    return not arr #if the stack is empty then string has valid parenthesis